File: src/mbs/presentation/interactor_factory.py
from abc import ABC, abstractmethod

from mbs.application.article.delete import DeleteArticle
from mbs.application.article.create import CreateArticle
from mbs.application.article.get import GetArticle
from mbs.application.article.get_range import GetArticleRange
from mbs.application.article.rate import RateArticle
from mbs.application.article.update import UpdateArticle

from mbs.application.comment.create import CreateComment
from mbs.application.comment.delete import DeleteComment
from mbs.application.comment.get import GetComment
from mbs.application.comment.get_by_article import GetArticleComments
from mbs.application.comment.update import UpdateComment

from mbs.application.file.confirm_article_file import ConfirmArticleFile
from mbs.application.file.create_article_file import CreateArticleFile
from mbs.application.file.delete_article_file import DeleteArticleFile
from mbs.application.file.get_article_file import GetArticleFile
from mbs.application.file.get_article_files import GetArticleFiles

from mbs.application.common.id_provider import IdProvider


class InteractorFactory(ABC):

    @abstractmethod
    def create_article(self, id_provider: IdProvider) -> CreateArticle:
        pass

    @abstractmethod
    def delete_article(self, id_provider: IdProvider) -> DeleteArticle:
        pass

    @abstractmethod
    def get_article(self, id_provider: IdProvider) -> GetArticle:
        pass

    @abstractmethod
    def get_article_range(self, id_provider: IdProvider) -> GetArticleRange:
        pass

    @abstractmethod
    def rate_article(self, id_provider: IdProvider) -> RateArticle:
        pass

    @abstractmethod
    def update_article(self, id_provider: IdProvider) -> UpdateArticle:
        pass

    @abstractmethod
    def create_comment(self, id_provider: IdProvider) -> CreateComment:
        pass

    @abstractmethod
    def delete_comment(self, id_provider: IdProvider) -> DeleteComment:
        pass

    @abstractmethod
    def get_comment(self, id_provider: IdProvider) -> GetComment:
        pass

    @abstractmethod
    def get_article_comments(self, id_provider: IdProvider) -> GetArticleComments:
        pass

    @abstractmethod
    def update_comment(self, id_provider: IdProvider) -> UpdateComment:
        pass

    @abstractmethod
    def create_article_file(self, id_provider: IdProvider) -> CreateArticleFile:
        pass

    @abstractmethod
    def delete_article_file(self, id_provider: IdProvider) -> DeleteArticleFile:
        pass

    @abstractmethod
    def confirm_article_file(self, id_provider: IdProvider) -> ConfirmArticleFile:
        pass

    @abstractmethod
    def get_article_files(self, id_provider: IdProvider) -> GetArticleFiles:
        pass

    @abstractmethod
    def get_article_file(self, id_provider: IdProvider) -> GetArticleFile:
        pass

File: src/presentation/interactor_factory.rs
use crate::application::{
    common::id_provider::IdProvider,
    article::{
        create::CreateArticle,
        delete::DeleteArticle,
        find::FindArticle,
        get::GetArticle,
        rate::RateArticle,
        update::UpdateArticle,
        file::{
            confirm::ConfirmArticleFile,
            create::CreateArticleFile,
            delete::DeleteArticleFile
        },
        tag::find::FindArticleTags
    },
    comment::{
        create::CreateComment,
        delete::DeleteComment,
        get::GetComment,
        get_tree::GetCommentsTree,
        rate::RateComment,
        update::UpdateComment
    }
};

pub trait InteractorFactory: Send + Sync {
    fn create_article<'interactor>(&'interactor self, id_provider: Box<dyn IdProvider>) -> CreateArticle<'interactor>;
    fn delete_article<'interactor>(&'interactor self, id_provider: Box<dyn IdProvider>) -> DeleteArticle<'interactor>;
    fn find_article<'interactor>(&'interactor self, id_provider: Box<dyn IdProvider>) -> FindArticle<'interactor>;
    fn get_article<'interactor>(&'interactor self, id_provider: Box<dyn IdProvider>) -> GetArticle<'interactor>;
    fn rate_article<'interactor>(&'interactor self, id_provider: Box<dyn IdProvider>) -> RateArticle<'interactor>;
    fn update_article<'interactor>(&'interactor self, id_provider: Box<dyn IdProvider>) -> UpdateArticle<'interactor>;
    fn create_article_file<'interactor>(&'interactor self, id_provider: Box<dyn IdProvider>) -> CreateArticleFile<'interactor>;
    fn confirm_article_file<'interactor>(&'interactor self, id_provider: Box<dyn IdProvider>) -> ConfirmArticleFile<'interactor>;
    fn delete_article_file<'interactor>(&'interactor self, id_provider: Box<dyn IdProvider>) -> DeleteArticleFile<'interactor>;
    fn find_article_tags<'interactor>(&'interactor self, id_provider: Box<dyn IdProvider>) -> FindArticleTags<'interactor>;
    // ---------------------------------------------------------------------------------------------
    fn create_comment<'interactor>(&'interactor self, id_provider: Box<dyn IdProvider>) -> CreateComment<'interactor>;
    fn delete_comment<'interactor>(&'interactor self, id_provider: Box<dyn IdProvider>) -> DeleteComment<'interactor>;
    fn get_comment<'interactor>(&'interactor self, id_provider: Box<dyn IdProvider>) -> GetComment<'interactor>;
    fn get_comments_tree<'interactor>(&'interactor self, id_provider: Box<dyn IdProvider>) -> GetCommentsTree<'interactor>;
    fn rate_comment<'interactor>(&'interactor self, id_provider: Box<dyn IdProvider>) -> RateComment<'interactor>;
    fn update_comment<'interactor>(&'interactor self, id_provider: Box<dyn IdProvider>) -> UpdateComment<'interactor>;
}